rfcs/0192-fr-copy-receipt-performance.md

```diff
@@ -0,0 +1,218 @@
+        var payload = JsonConvert.DeserializeObject<CopyPayload>(Encoding.UTF8.GetString(Utilities.FromBase64urlString(jwt[1])));
+        if (payload.CopiedReceiptReference == cbPreviousReceiptReference)
+        {
+            count++;
+        }
+    }
+    return count;
+}
+```
+
+> ***Note:** We will release this out first as an intermediate change to provide some initial improvements.*
+
+## Option 2: Search only for the latest CopyReceipt or receipt with the same `cbReceiptReference`
+
+An other alternative way of doing this would be to keep the current implementation and improve it.
+
+One way to do that would be to not count every CopyReceipt in the `ftJournalFR` table but only find the latest one and increase its count by one.
+
+If we find a non copy receipt with the same `cbReceiptReference` as the `cbPreviousReceiptReference` we can return 0 directly.
+
+```cs
+private async Task<int> GetCountOfExistingCopiesAsync(string cbPreviousReceiptReference)
+{
+    await foreach (var copyJournal in _journalFRRepository.GetProcessedCopyReceiptsDescAsync())
+    {
+        var queueItem = await _queueItemRepository.GetAsync(copyJournal.ftQueueItemId);
+        var response = queueItem?.response != null ? JsonConvert.DeserializeObject<ReceiptResponse>(queueItem.response) : null;
+        if (response != null)
+        {
+            var duplicata = response.ftSignatures.FirstOrDefault(x => x.Caption == "Duplicata" && x.Data.EndsWith($"Duplicata de {cbPreviousReceiptReference}"));
+
+            if (duplicata != null)
+            {
+                return int.Parse(duplicata.Data.Split('.')[0]);
+            }
+
+            if(response.cbReceiptReference == cbPreviousReceiptReference) {
+                return 0;
+            }
+        }
+    }
+    return 0;
+}
+```
+
+In the worst case scenario this would be as slow as the current implementation but in the best case scenario only need one database query.
+
+### Open questions
+
+* Can we implement a `GetProcessedCopyReceiptsDescAsync` method that gets us the CopyReceipts in descending order.
+  Can we sort by row number or `TimeStamp`?
+* Taking non copy receipts into account would change the behaviour of the middleware in the case of multiple different receipts with the same receipt reference. Is this acceptable?
+* Is the performance of this solution in a real world scenario enough?
+  And do we have clients which have a lot of CopyReceipts and need copies of older receipts where this would lean towards worst case performance?
+
+# Unresolved questions
+
+## Initialization
+
+Can we populate the table asynchronously?
+This would mean we could start the Queue (and process non CopyReceipts) without waiting for the population of the `ftJournalFRCopyPayload` table to finish.
+
+However this would mean we would have to either fail CopyReceipts until the process of population is complete or wait for the population to finish before we start processing the CopyReceipt.
```

Comment:
We'd need to test how long this takes i guess, IMO everything up to 25sec startup time (one-shot) would be okayish for existing implementations (given that the max startup time of a Windows service is 30 seconds)